src/ucp/proto/proto_common.c

```diff
@@ -707,11 +759,24 @@ void ucp_proto_request_bcopy_abort(ucp_request_t *request, ucs_status_t status)
     ucp_request_complete_send(request, status);
 }
 
+void ucp_proto_request_bcopy_clean(ucp_request_t *request)
+{
+    ucs_assert(ucp_datatype_iter_is_begin(&request->send.state.dt_iter));
+    ucp_datatype_iter_cleanup(&request->send.state.dt_iter, UCP_DT_MASK_ALL);
+    request->flags &= ~UCP_REQUEST_FLAG_PROTO_INITIALIZED;
+}
+
 void ucp_proto_request_zcopy_abort(ucp_request_t *request, ucs_status_t status)
 {
     ucp_invoke_uct_completion(&request->send.state.uct_comp, status);
 }
 
+void ucp_proto_request_zcopy_clean(ucp_request_t *request)
+{
+    ucs_assert(ucp_datatype_iter_is_begin(&request->send.state.dt_iter));
```

Comment:
maybe have helper function for this assertion, that also prints request and it's current offset (with ucs_assertv)